06_puppy/madhavansu/main.go

```diff
@@ -0,0 +1,114 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"sync"
+)
+
+var out io.Writer = os.Stdout
+
+//Sync store
+type syncStore struct {
+	sync.Mutex
+	sync.Map
+}
+
+func newSyncStore() *syncStore {
+	return &syncStore{}
+}
+
+func (s *syncStore) createPuppy(in Puppy) {
+	s.Store(in.id, in)
+}
+
+func (s *syncStore) readPuppy(id uint) Puppy {
+	pd, ok := s.Load(id)
+	if !ok {
+		fmt.Printf("No puppy exists\n")
+	}
+	p, _ := pd.(Puppy)
+	return p
+}
+
+func (s *syncStore) updatePuppy(id uint, in Puppy) {
+	s.Store(in.id, in)
+}
+
+func (s *syncStore) deletePuppy(id uint) {
+	s.Delete(id)
+}
+
+// Map Store
+type mapStore struct {
+	ms    map[uint]Puppy
+	mapID uint
+}
+
+func newMapStore() *mapStore {
+	return &mapStore{make(map[uint]Puppy), 0}
+}
+
+func (m *mapStore) createPuppy(in Puppy) uint {
+	m.mapID++
+	in.id = m.mapID
+	m.ms[in.id] = in
+	return m.mapID
+}
+
+func (m *mapStore) readPuppy(id uint) Puppy {
+	p, ok := m.ms[id]
+	if !ok {
+		fmt.Printf("No puppy exists\n")
+	}
+	return p
+}
+
+func (m *mapStore) updatePuppy(id uint, in Puppy) {
+	m.ms[id] = in
+}
+
+func (m *mapStore) deletePuppy(id uint) {
+	_, ok := m.ms[id]
+	if ok {
+		delete(m.ms, id)
+	}
+}
+
+func main() {
+	// Sync storer
+	var s Storer = newSyncStore()
+	fmt.Fprintln(out, "~~~~~~~~~~")
+	fmt.Fprintln(out, "Sync Store")
+	fmt.Fprintln(out, "~~~~~~~~~~")
+	// Create/Read puppy
+	s.createPuppy(Puppy{101, "Poodle", "red", 18000})
+	fmt.Fprintln(out, "101 : ", s.readPuppy(101))
+	// Delete puppy
+	s.createPuppy(Puppy{102, "Bulldog", "brownish", 9999})
+	s.deletePuppy(102)
+	fmt.Fprintln(out, "102 : ", s.readPuppy(102))
+	// Update puppy
+	s.createPuppy(Puppy{103, "Labrador Retriever", "purple", 987})
+	s.updatePuppy(103, Puppy{103, "German Shepherd", "red", 4533})
+	fmt.Fprintln(out, "103 : ", s.readPuppy(103))
+
+	// Map storer
+	var m StorerMap = newMapStore()
+	var puppyID uint
+	fmt.Fprintln(out, "~~~~~~~~~")
+	fmt.Fprintln(out, "Map Store")
+	fmt.Fprintln(out, "~~~~~~~~~")
+	// Create/Read puppy
+	puppyID = m.createPuppy(Puppy{104, "Pug", "brown", 0.91})
+	fmt.Fprintln(out, puppyID, " : ", m.readPuppy(puppyID))
+	// Delete puppy
+	puppyID = m.createPuppy(Puppy{105, "Beagle", "yellowish", 1233})
+	m.deletePuppy(puppyID)
+	fmt.Fprintln(out, puppyID, " : ", m.readPuppy(puppyID))
+	// Update puppy
+	puppyID = m.createPuppy(Puppy{106, "Boxer", "black", 45000.98})
+	m.updatePuppy(puppyID, Puppy{puppyID, "Beagle", "brown", 0.91})
+	fmt.Fprint(out, puppyID, " :  ", m.readPuppy(puppyID))
```

Comment:
use `fmt.Fprintln` here so it properly outputs a newline at the end. For reasons I don't understand, this also adds the extra space so the colons line up, so also remove that extra space after the colon that you have that lines up the values. Don't forget to update your tests.

Reply:
use `FPrintln`, otherwise you get yucky terminal output.